### ISSUE TITLE
receive: v0.18.0 memory leak (v0.16.0 regression)

### ISSUE THREAD

**Issue by sepich** [thanos-io/thanos#3726]
**Thanos, Prometheus and Golang version used**:
thanosio/thanos:v0.17.2

**Object Storage Provider**:
GCS

**What happened**:
I'm trying to upgrade from v0.16.0 to v0.17.2 and see that thanos-receive memory is "leaking":
![image](https://user-images.githubusercontent.com/1803804/104817758-4428cb80-5834-11eb-8f40-d281b4fd986f.png)
Here are 3 thanos-receive pods in a hashring with equal load, red and blue lines are v0.16.0. At 18:30 i'm restarting thanos-receive-0 (orange line) as v0.17.2, then at 19:33 restart it back as v0.16.0.
GC load profile also differs between versions:
![image](https://user-images.githubusercontent.com/1803804/104817776-6cb0c580-5834-11eb-832d-abb5ee9bb05a.png)

Related:
https://github.com/thanos-io/thanos/issues/3265

**What you expected to happen**:
Stable memory usage.

**How to reproduce it (as minimally and precisely as possible)**:
Only reproducible on production with 80k samples/s per thanos-receive pod.

**Full logs to relevant components**:
Attaching pprof heap.zip, right before second restart.
[heap.zip](https://github.com/thanos-io/thanos/files/5824768/heap.zip)

**Comment by sepich** [comment]
Rechecked on:
```
thanos, version 0.18.0 (branch: HEAD, revision: 60d45a02d46858a38013283b578017a171cf7b82)
  build user:       circleci@8ddf80c1eb30
  build date:       20210127-12:29:07
  go version:       go1.15.7
  platform:         linux/amd64
```
Result still the same (v0.16.0 before 17:43):
![image](https://user-images.githubusercontent.com/1803804/107154530-7ab1bc00-6984-11eb-8e7b-095b3a75b9e5.png)
![image](https://user-images.githubusercontent.com/1803804/107154547-87ceab00-6984-11eb-9f52-d01e83c540fc.png)
Attaching pprof for v0.18.0 if needed:
[heap.zip](https://github.com/thanos-io/thanos/files/5939590/heap.zip)

**Comment by kakkoyun** [comment]
Sorry for the late response @sepich, is this still an issue with the latest Thanos version?

**Comment by sepich** [comment]
Result for latest v0.18.0 is above.
Anyway i've retested with `thanosio/thanos:master-2021-02-11-7b09e30c` and memory is leaking to 24Gb even faster.

**Comment by kakkoyun** [comment]
Thanks again. We're going to dedicate something to investigate this. 
We had a similar issue but I thought we had fixed it. Anyways we'll check it out. Feel free to give us a hand, so help wanted.

cc @squat @bwplotka

**Comment by kakkoyun** [comment]
Possible duplicate of #3471

**Comment by luizrojo** [comment]
I am also facing this issue and same behavior, v0.16.0 fixes the possible leak.

As we can see below

<img width="2524" alt="Screen Shot 2021-02-23 at 10 49 37" src="https://user-images.githubusercontent.com/2788953/108852844-d905ae00-75c4-11eb-8af3-04f8efd49bea.png">

Until 9h40 all nodes were running v0.18.0, at this time I had all receive nodes restarted.

At 9h45 all nodes were back up, still on v0.18.0, and we can see a rapid memory increase.

At 10h17 I had 2 nodes downgraded to v0.17.2 and v0.16.0. (ip-10-184-125-10 to v0.17.2 and ip-10-184-125-13 to v0.16.0)

At 10h35 the v0.17.2 node starts to increase the memory usage, following the v0.18.0 nodes behavior, but the v0.16.0 keeps the memory usage stable.

Here we can see the memory heap usage

<img width="2548" alt="Screen Shot 2021-02-23 at 10 50 27" src="https://user-images.githubusercontent.com/2788953/108853022-005c7b00-75c5-11eb-929c-14ba573c906f.png">

**Comment by luizrojo** [comment]
@bwplotka I believe something changed between 0.16.0 and 0.17.2 and then apparently got worse on 0.18.0.

On 0.17.2 we can see it takes longer to start building the memory usage, but it does start to increase, following the same behavior as 0.18.0.

I'll update one of the instance to the v0.19.0-rc.0 and get back to you with some more info.

Right now I got all nodes on v0.16.0 and memory is as stable as it can get:

<img width="2545" alt="Screen Shot 2021-02-23 at 12 12 23" src="https://user-images.githubusercontent.com/2788953/108863991-892ce400-75d0-11eb-9482-c04fde6af1ba.png">

<img width="2550" alt="Screen Shot 2021-02-23 at 12 12 55" src="https://user-images.githubusercontent.com/2788953/108864008-8df19800-75d0-11eb-88f3-042627ee44f9.png">

At 11h18 I had all instances downgraded and at 11h34 got metrics ingestion back on

**Comment by bwplotka** [comment]
I would start bisecting the commits between 0.17.2  🤗  that would be helpful.

**Comment by luizrojo** [comment]
I just ran a couple more tests and did not have to go very far on the image tags to notice a pattern change on memory consumption.

Here is a screenshot of the memory graphs, the green line is the test instance and the yellow line is the instance running 0.16.0.

The first 3 big slopes are version 0.17.0, 0.17.2 and 0.18.0 consecutively 

Starting at 15h49 im using `master-2020-10-26-8447f621` the first image tag after 0.16.0.

The pattern changes, and it looks like after every GC, the consumption increases a little. 

<img width="2552" alt="Screen Shot 2021-02-23 at 16 19 09" src="https://user-images.githubusercontent.com/2788953/108895949-02d5c980-75f3-11eb-8fc6-68c43a35865f.png">

**Comment by bwplotka** [comment]
There are cool findings made by @svenwltr:
 https://github.com/thanos-io/thanos/issues/3471#issuecomment-773190157

**Comment by jmichalek132** [comment]
We are experiencing the same issue (with both v0.17.2 and v0.18.0), but the downgrade to v0.16.0 seems to help as suggested.

![Screenshot_2021-03-01 Thanos Receive - Grafana](https://user-images.githubusercontent.com/7171250/109516234-3f4c6e00-7aa8-11eb-86cc-774ac3a05c9f.png)

The fall in memory usage between 13:20 and 14:40 is caused by the instances beeing oomkilled.
The fall in memory usage between 15:20 and 14:50 is caused by the rollout of v0.16.0. After that the memory usage is lower and stable.

**Comment by dhohengassner** [comment]
We are facing the same issue. Glad this issue exists.  
Downgrade to 0.16.0 stabilized our system as well :+1: 

Looking forward to a fix!

**Comment by jmichalek132** [comment]
Tried with ` v0.19.0-rc.0 ` to see if the issue is still there, looks like it is. (Upgraded from 0.16.0 to the rc at 14:30)
![Screenshot_2021-03-02 Thanos Receive - Grafana](https://user-images.githubusercontent.com/7171250/109667557-8bfb7c00-7b70-11eb-8e85-08d5f208a960.png)

**Comment by bwplotka** [comment]
Thanks a lot!

Anyone can help but I would love to find out what's wrong this week, ideally before 0.19.0 but let's see.

We got some profile but let's see if this is helpful:

```
(pprof) top 10
Showing nodes accounting for 421.25GB, 96.30% of 437.43GB total
Dropped 572 nodes (cum <= 2.19GB)
Showing top 10 nodes out of 47
      flat  flat%   sum%        cum   cum%
  199.27GB 45.55% 45.55%   199.27GB 45.55%  github.com/thanos-io/thanos/pkg/store/storepb/prompb.(*TimeSeries).Unmarshal
   80.06GB 18.30% 63.86%    80.06GB 18.30%  github.com/golang/snappy.Decode
   75.49GB 17.26% 81.11%    77.47GB 17.71%  github.com/thanos-io/thanos/pkg/receive.(*Writer).Write
   25.18GB  5.76% 86.87%    25.18GB  5.76%  bytes.makeSlice
   18.15GB  4.15% 91.02%    18.64GB  4.26%  github.com/thanos-io/thanos/pkg/receive.(*Handler).forward
   18.07GB  4.13% 95.15%   217.34GB 49.69%  github.com/thanos-io/thanos/pkg/store/storepb/prompb.(*WriteRequest).Unmarshal
    2.71GB  0.62% 95.77%     2.71GB  0.62%  github.com/prometheus/prometheus/tsdb/encoding.(*Decbuf).UvarintStr (inline)
    1.97GB  0.45% 96.22%     4.62GB  1.06%  github.com/prometheus/prometheus/tsdb/record.(*Decoder).Series
    0.33GB 0.076% 96.29%     2.87GB  0.66%  github.com/prometheus/prometheus/tsdb.(*Head).loadWAL
    0.03GB 0.0064% 96.30%     2.32GB  0.53%  github.com/prometheus/prometheus/tsdb.(*Head).loadWAL.func6
```

Ideally we pin point commit which introduced the regression 🤗 The problem will be if that's TSDB update (it probably is).

**Comment by jmichalek132** [comment]
Profiles captured via conprof:
```

github.com/thanos-io/thanos/pkg/store/storepb/prompb.(*TimeSeries).Unmarshal

/home/circleci/project/pkg/store/storepb/prompb/types.pb.go

  Total:     22.64TB    22.64TB (flat, cum) 46.00%

github.com/thanos-io/thanos/pkg/receive.(*Writer).Write

/home/circleci/project/pkg/receive/writer.go

  Total:      6.77TB     6.88TB (flat, cum) 13.98%

google.golang.org/grpc.(*parser).recvMsg

/home/circleci/go/pkg/mod/google.golang.org/grpc@v1.29.1/rpc_util.go

  Total:      4.77TB     4.77TB (flat, cum)  9.70%

github.com/thanos-io/thanos/pkg/store/storepb.(*WriteRequest).Marshal

/home/circleci/project/pkg/store/storepb/rpc.pb.go

  Total:      4.77TB     4.77TB (flat, cum)  9.69%

github.com/golang/snappy.Decode

/home/circleci/go/pkg/mod/github.com/golang/snappy@v0.0.2/decode.go

  Total:      1.79TB     1.79TB (flat, cum)  3.63%

github.com/thanos-io/thanos/pkg/receive.(*Handler).forward

/home/circleci/project/pkg/receive/handler.go

  Total:      1.72TB     3.31TB (flat, cum)  6.72%

google.golang.org/grpc/internal/transport.(*http2Client).Write

/home/circleci/go/pkg/mod/google.golang.org/grpc@v1.29.1/internal/transport/http2_client.go

  Total:      1.60TB     1.60TB (flat, cum)  3.25%

github.com/thanos-io/thanos/pkg/store/storepb.(*WriteRequest).Unmarshal

/home/circleci/project/pkg/store/storepb/rpc.pb.go

  Total:      1.18TB    17.34TB (flat, cum) 35.22%

internal/reflectlite.Swapper

/usr/local/go/src/internal/reflectlite/swapper.go

  Total:    937.76GB   937.76GB (flat, cum)  1.86%

github.com/thanos-io/thanos/pkg/receive.hash

/home/circleci/project/pkg/receive/hashring.go

  Total:    603.91GB     1.51TB (flat, cum)  3.06%

bytes.makeSlice

/usr/local/go/src/bytes/buffer.go

  Total:    459.36GB   459.36GB (flat, cum)  0.91%

github.com/thanos-io/thanos/pkg/store/storepb/prompb.(*WriteRequest).Unmarshal

/home/circleci/project/pkg/store/storepb/prompb/remote.pb.go

  Total:    325.43GB     6.80TB (flat, cum) 13.82%

github.com/prometheus/prometheus/tsdb/record.(*Decoder).Series

/home/circleci/go/pkg/mod/github.com/prometheus/prometheus@v1.8.2-0.20201119142752-3ad25a6dc3d9/tsdb/record/record.go

  Total:    138.39GB   138.39GB (flat, cum)  0.27%

google.golang.org/grpc/internal/transport.(*decodeState).processHeaderField

/home/circleci/go/pkg/mod/google.golang.org/grpc@v1.29.1/internal/transport/http_util.go

  Total:     87.70GB    87.70GB (flat, cum)  0.17%

google.golang.org/grpc/internal/transport.(*http2Client).newStream

/home/circleci/go/pkg/mod/google.golang.org/grpc@v1.29.1/internal/transport/http2_client.go

  Total:     87.38GB   111.23GB (flat, cum)  0.22%

golang.org/x/net/http2.(*Framer).readMetaFrame.func1

/home/circleci/go/pkg/mod/golang.org/x/net@v0.0.0-20201110031124-69a78807bb2b/http2/frame.go

  Total:     85.90GB    85.90GB (flat, cum)  0.17%

google.golang.org/grpc/internal/transport.(*http2Client).createHeaderFields

/home/circleci/go/pkg/mod/google.golang.org/grpc@v1.29.1/internal/transport/http2_client.go

  Total:     85.67GB    95.15GB (flat, cum)  0.19%

context.WithValue

/usr/local/go/src/context/context.go

  Total:     85.18GB    85.18GB (flat, cum)  0.17%

github.com/prometheus/prometheus/tsdb.(*blockBaseSeriesSet).Next

/home/circleci/go/pkg/mod/github.com/prometheus/prometheus@v1.8.2-0.20201119142752-3ad25a6dc3d9/tsdb/querier.go

  Total:     75.53GB    75.58GB (flat, cum)  0.15%

google.golang.org/grpc/internal/transport.(*http2Server).operateHeaders

/home/circleci/go/pkg/mod/google.golang.org/grpc@v1.29.1/internal/transport/http2_server.go

  Total:     72.84GB   174.99GB (flat, cum)  0.35%

google.golang.org/grpc.newClientStream

/home/circleci/go/pkg/mod/google.golang.org/grpc@v1.29.1/stream.go

  Total:     62.68GB   332.85GB (flat, cum)  0.66%

github.com/go-kit/kit/log.With

/home/circleci/go/pkg/mod/github.com/go-kit/kit@v0.10.0/log/log.go

  Total:     46.79GB    46.79GB (flat, cum) 0.093%

github.com/prometheus/prometheus/tsdb/index.(*Writer).writePostingsToTmpFiles

/home/circleci/go/pkg/mod/github.com/prometheus/prometheus@v1.8.2-0.20201119142752-3ad25a6dc3d9/tsdb/index/index.go

  Total:     43.93GB    45.88GB (flat, cum) 0.091%

golang.org/x/net/http2.(*Framer).readMetaFrame

/home/circleci/go/pkg/mod/golang.org/x/net@v0.0.0-20201110031124-69a78807bb2b/http2/frame.go

  Total:     40.48GB   127.44GB (flat, cum)  0.25%

google.golang.org/grpc/internal/transport.(*controlBuffer).executeAndPut

/home/circleci/go/pkg/mod/google.golang.org/grpc@v1.29.1/internal/transport/controlbuf.go

  Total:     36.12GB    39.32GB (flat, cum) 0.078%

context.(*cancelCtx).Done

/usr/local/go/src/context/context.go

  Total:     33.51GB    33.51GB (flat, cum) 0.066%

context.WithDeadline

/usr/local/go/src/context/context.go

  Total:     31.74GB    51.55GB (flat, cum)   0.1%

google.golang.org/grpc/internal/transport.newWriteQuota

/home/circleci/go/pkg/mod/google.golang.org/grpc@v1.29.1/internal/transport/flowcontrol.go

  Total:     28.73GB    28.73GB (flat, cum) 0.057%

github.com/thanos-io/thanos/pkg/receive.(*Handler).replicate

/home/circleci/project/pkg/receive/handler.go

  Total:     23.55GB    72.43GB (flat, cum)  0.14%

github.com/thanos-io/thanos/pkg/tracing.StartSpan

/home/circleci/project/pkg/tracing/tracing.go

  Total:     22.55GB    45.18GB (flat, cum)  0.09%

github.com/prometheus/prometheus/tsdb.(*Head).appender

/home/circleci/go/pkg/mod/github.com/prometheus/prometheus@v1.8.2-0.20201119142752-3ad25a6dc3d9/tsdb/head.go

  Total:     22.28GB    22.30GB (flat, cum) 0.044%

github.com/prometheus/prometheus/tsdb/index.(*MemPostings).addFor

/home/circleci/go/pkg/mod/github.com/prometheus/prometheus@v1.8.2-0.20201119142752-3ad25a6dc3d9/tsdb/index/postings.go

  Total:     20.98GB    20.98GB (flat, cum) 0.042%

github.com/prometheus/prometheus/tsdb.(*blockBaseSeriesSet).Next.func1

/home/circleci/go/pkg/mod/github.com/prometheus/prometheus@v1.8.2-0.20201119142752-3ad25a6dc3d9/tsdb/querier.go

  Total:     19.84GB    19.84GB (flat, cum) 0.039%

time.AfterFunc

/usr/local/go/src/time/sleep.go

  Total:     19.81GB    19.81GB (flat, cum) 0.039%

github.com/prometheus/prometheus/tsdb.(*memSeries).cutNewHeadChunk

/home/circleci/go/pkg/mod/github.com/prometheus/prometheus@v1.8.2-0.20201119142752-3ad25a6dc3d9/tsdb/head.go

  Total:     19.62GB    49.24GB (flat, cum) 0.098%

github.com/thanos-io/thanos/pkg/receive.(*Handler).fanoutForward

/home/circleci/project/pkg/receive/handler.go

  Total:     18.12GB   123.33GB (flat, cum)  0.24%

google.golang.org/grpc/internal/transport.(*http2Server).writeHeaderLocked

/home/circleci/go/pkg/mod/google.golang.org/grpc@v1.29.1/internal/transport/http2_server.go

  Total:     17.57GB    20.73GB (flat, cum) 0.041%

google.golang.org/grpc/internal/transport.(*http2Server).WriteStatus

/home/circleci/go/pkg/mod/google.golang.org/grpc@v1.29.1/internal/transport/http2_server.go

  Total:     17.49GB    37.89GB (flat, cum) 0.075%

github.com/grpc-ecosystem/go-grpc-middleware/tracing/opentracing.newClientSpanFromContext

/home/circleci/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.1.0/tracing/opentracing/client_interceptors.go

  Total:     17.46GB    36.64GB (flat, cum) 0.073%

github.com/prometheus/prometheus/tsdb/index.(*MemPostings).Delete

/home/circleci/go/pkg/mod/github.com/prometheus/prometheus@v1.8.2-0.20201119142752-3ad25a6dc3d9/tsdb/index/postings.go

  Total:     16.26GB    16.26GB (flat, cum) 0.032%

google.golang.org/grpc/internal/transport.(*http2Client).NewStream

/home/circleci/go/pkg/mod/google.golang.org/grpc@v1.29.1/internal/transport/http2_client.go

  Total:     15.73GB   233.06GB (flat, cum)  0.46%

github.com/prometheus/prometheus/tsdb.(*Head).processWALSamples

/home/circleci/go/pkg/mod/github.com/prometheus/prometheus@v1.8.2-0.20201119142752-3ad25a6dc3d9/tsdb/head.go

  Total:     15.57GB    18.77GB (flat, cum) 0.037%

golang.org/x/net/http2.parseHeadersFrame

/home/circleci/go/pkg/mod/golang.org/x/net@v0.0.0-20201110031124-69a78807bb2b/http2/frame.go

  Total:     14.25GB    14.25GB (flat, cum) 0.028%

google.golang.org/grpc/internal/transport.(*recvBuffer).put

/home/circleci/go/pkg/mod/google.golang.org/grpc@v1.29.1/internal/transport/transport.go

  Total:     13.98GB    13.98GB (flat, cum) 0.028%

google.golang.org/grpc.(*Server).processUnaryRPC

/home/circleci/go/pkg/mod/google.golang.org/grpc@v1.29.1/server.go

  Total:     13.16GB    24.68TB (flat, cum) 50.12%

google.golang.org/grpc.(*clientStream).newAttemptLocked

/home/circleci/go/pkg/mod/google.golang.org/grpc@v1.29.1/stream.go

  Total:     12.72GB    12.72GB (flat, cum) 0.025%

github.com/grpc-ecosystem/go-grpc-prometheus.newClientReporter

/home/circleci/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-prometheus@v1.2.0/client_reporter.go

  Total:     12.72GB    12.72GB (flat, cum) 0.025%

github.com/grpc-ecosystem/go-grpc-prometheus.newServerReporter

/home/circleci/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-prometheus@v1.2.0/server_reporter.go

  Total:     12.67GB    12.67GB (flat, cum) 0.025%

google.golang.org/grpc.(*clientStream).SendMsg

/home/circleci/go/pkg/mod/google.golang.org/grpc@v1.29.1/stream.go

  Total:     12.66GB     6.39TB (flat, cum) 12.97%

github.com/prometheus/prometheus/tsdb/chunkenc.(*XORChunk).iterator

/home/circleci/go/pkg/mod/github.com/prometheus/prometheus@v1.8.2-0.20201119142752-3ad25a6dc3d9/tsdb/chunkenc/xor.go

  Total:     12.15GB    12.15GB (flat, cum) 0.024%

```

```

github.com/thanos-io/thanos/pkg/store/storepb/prompb.(*TimeSeries).Unmarshal

/home/circleci/project/pkg/store/storepb/prompb/types.pb.go

  Total:      2.11TB     2.11TB (flat, cum) 45.68%

github.com/thanos-io/thanos/pkg/receive.(*Writer).Write

/home/circleci/project/pkg/receive/writer.go

  Total:    644.39GB   655.40GB (flat, cum) 13.88%

github.com/thanos-io/thanos/pkg/store/storepb.(*WriteRequest).Marshal

/home/circleci/project/pkg/store/storepb/rpc.pb.go

  Total:    453.94GB   453.94GB (flat, cum)  9.61%

google.golang.org/grpc.(*parser).recvMsg

/home/circleci/go/pkg/mod/google.golang.org/grpc@v1.29.1/rpc_util.go

  Total:    452.94GB   452.99GB (flat, cum)  9.59%

github.com/golang/snappy.Decode

/home/circleci/go/pkg/mod/github.com/golang/snappy@v0.0.2/decode.go

  Total:    172.04GB   172.04GB (flat, cum)  3.64%

github.com/thanos-io/thanos/pkg/receive.(*Handler).forward

/home/circleci/project/pkg/receive/handler.go

  Total:    163.47GB   314.08GB (flat, cum)  6.65%

google.golang.org/grpc/internal/transport.(*http2Client).Write

/home/circleci/go/pkg/mod/google.golang.org/grpc@v1.29.1/internal/transport/http2_client.go

  Total:    151.62GB   151.90GB (flat, cum)  3.22%

github.com/thanos-io/thanos/pkg/store/storepb.(*WriteRequest).Unmarshal

/home/circleci/project/pkg/store/storepb/rpc.pb.go

  Total:    111.61GB     1.61TB (flat, cum) 34.83%

internal/reflectlite.Swapper

/usr/local/go/src/internal/reflectlite/swapper.go

  Total:     87.29GB    87.29GB (flat, cum)  1.85%

github.com/thanos-io/thanos/pkg/receive.hash

/home/circleci/project/pkg/receive/hashring.go

  Total:     55.46GB   142.74GB (flat, cum)  3.02%

bytes.makeSlice

/usr/local/go/src/bytes/buffer.go

  Total:     44.68GB    44.68GB (flat, cum)  0.95%

github.com/thanos-io/thanos/pkg/store/storepb/prompb.(*WriteRequest).Unmarshal

/home/circleci/project/pkg/store/storepb/prompb/remote.pb.go

  Total:     30.47GB   654.21GB (flat, cum) 13.86%

github.com/prometheus/prometheus/tsdb/record.(*Decoder).Series

/home/circleci/go/pkg/mod/github.com/prometheus/prometheus@v1.8.2-0.20201119142752-3ad25a6dc3d9/tsdb/record/record.go

  Total:     19.57GB    19.57GB (flat, cum)  0.41%

github.com/prometheus/prometheus/tsdb.(*Head).processWALSamples

/home/circleci/go/pkg/mod/github.com/prometheus/prometheus@v1.8.2-0.20201119142752-3ad25a6dc3d9/tsdb/head.go

  Total:     15.57GB    18.77GB (flat, cum)   0.4%

google.golang.org/grpc/internal/transport.(*decodeState).processHeaderField

/home/circleci/go/pkg/mod/google.golang.org/grpc@v1.29.1/internal/transport/http_util.go

  Total:      8.16GB     8.16GB (flat, cum)  0.17%

google.golang.org/grpc/internal/transport.(*http2Client).newStream

/home/circleci/go/pkg/mod/google.golang.org/grpc@v1.29.1/internal/transport/http2_client.go

  Total:      8.02GB    10.23GB (flat, cum)  0.22%

google.golang.org/grpc/internal/transport.(*http2Client).createHeaderFields

/home/circleci/go/pkg/mod/google.golang.org/grpc@v1.29.1/internal/transport/http2_client.go

  Total:      8.02GB     8.89GB (flat, cum)  0.19%

github.com/prometheus/prometheus/tsdb/index.(*MemPostings).addFor

/home/circleci/go/pkg/mod/github.com/prometheus/prometheus@v1.8.2-0.20201119142752-3ad25a6dc3d9/tsdb/index/postings.go

  Total:      7.96GB     7.96GB (flat, cum)  0.17%

golang.org/x/net/http2.(*Framer).readMetaFrame.func1

/home/circleci/go/pkg/mod/golang.org/x/net@v0.0.0-20201110031124-69a78807bb2b/http2/frame.go

  Total:      7.93GB     7.93GB (flat, cum)  0.17%

context.WithValue

/usr/local/go/src/context/context.go

  Total:      7.93GB     7.93GB (flat, cum)  0.17%

google.golang.org/grpc/internal/transport.(*http2Server).operateHeaders

/home/circleci/go/pkg/mod/google.golang.org/grpc@v1.29.1/internal/transport/http2_server.go

  Total:      6.74GB    16.29GB (flat, cum)  0.34%

github.com/prometheus/prometheus/tsdb.(*blockBaseSeriesSet).Next

/home/circleci/go/pkg/mod/github.com/prometheus/prometheus@v1.8.2-0.20201119142752-3ad25a6dc3d9/tsdb/querier.go

  Total:         6GB     6.01GB (flat, cum)  0.13%

google.golang.org/grpc.newClientStream

/home/circleci/go/pkg/mod/google.golang.org/grpc@v1.29.1/stream.go

  Total:      5.82GB    30.78GB (flat, cum)  0.65%

github.com/go-kit/kit/log.With

/home/circleci/go/pkg/mod/github.com/go-kit/kit@v0.10.0/log/log.go

  Total:      4.26GB     4.26GB (flat, cum)  0.09%

golang.org/x/net/http2.(*Framer).readMetaFrame

/home/circleci/go/pkg/mod/golang.org/x/net@v0.0.0-20201110031124-69a78807bb2b/http2/frame.go

  Total:      3.77GB    11.81GB (flat, cum)  0.25%

github.com/prometheus/prometheus/tsdb/index.(*Writer).writePostingsToTmpFiles

/home/circleci/go/pkg/mod/github.com/prometheus/prometheus@v1.8.2-0.20201119142752-3ad25a6dc3d9/tsdb/index/index.go

  Total:      3.47GB     3.64GB (flat, cum) 0.077%

google.golang.org/grpc/internal/transport.(*controlBuffer).executeAndPut

/home/circleci/go/pkg/mod/google.golang.org/grpc@v1.29.1/internal/transport/controlbuf.go

  Total:      3.15GB     3.45GB (flat, cum) 0.073%

context.(*cancelCtx).Done

/usr/local/go/src/context/context.go

  Total:      3.14GB     3.14GB (flat, cum) 0.066%

context.WithDeadline

/usr/local/go/src/context/context.go

  Total:      2.98GB     4.85GB (flat, cum)   0.1%

github.com/prometheus/prometheus/tsdb.(*memSeries).cutNewHeadChunk

/home/circleci/go/pkg/mod/github.com/prometheus/prometheus@v1.8.2-0.20201119142752-3ad25a6dc3d9/tsdb/head.go

  Total:      2.95GB     6.55GB (flat, cum)  0.14%

google.golang.org/grpc/internal/transport.newWriteQuota

/home/circleci/go/pkg/mod/google.golang.org/grpc@v1.29.1/internal/transport/flowcontrol.go

  Total:      2.69GB     2.69GB (flat, cum) 0.057%

github.com/prometheus/prometheus/tsdb.(*Head).appender

/home/circleci/go/pkg/mod/github.com/prometheus/prometheus@v1.8.2-0.20201119142752-3ad25a6dc3d9/tsdb/head.go

  Total:      2.50GB     2.50GB (flat, cum) 0.053%

github.com/prometheus/prometheus/tsdb.(*Head).getOrCreateWithID

/home/circleci/go/pkg/mod/github.com/prometheus/prometheus@v1.8.2-0.20201119142752-3ad25a6dc3d9/tsdb/head.go

  Total:      2.49GB    12.46GB (flat, cum)  0.26%

github.com/prometheus/prometheus/tsdb/index.(*MemPostings).Delete

/home/circleci/go/pkg/mod/github.com/prometheus/prometheus@v1.8.2-0.20201119142752-3ad25a6dc3d9/tsdb/index/postings.go

  Total:      2.47GB     2.47GB (flat, cum) 0.052%

github.com/thanos-io/thanos/pkg/receive.(*Handler).replicate

/home/circleci/project/pkg/receive/handler.go

  Total:      2.26GB     6.79GB (flat, cum)  0.14%

github.com/thanos-io/thanos/pkg/tracing.StartSpan

/home/circleci/project/pkg/tracing/tracing.go

  Total:      2.07GB     4.19GB (flat, cum) 0.089%

time.AfterFunc

/usr/local/go/src/time/sleep.go

  Total:      1.86GB     1.87GB (flat, cum)  0.04%

github.com/prometheus/prometheus/tsdb/chunkenc.(*bstream).writeBits

/home/circleci/go/pkg/mod/github.com/prometheus/prometheus@v1.8.2-0.20201119142752-3ad25a6dc3d9/tsdb/chunkenc/bstream.go

  Total:      1.75GB     1.75GB (flat, cum) 0.037%

github.com/prometheus/prometheus/tsdb/chunkenc.(*XORChunk).iterator

/home/circleci/go/pkg/mod/github.com/prometheus/prometheus@v1.8.2-0.20201119142752-3ad25a6dc3d9/tsdb/chunkenc/xor.go

  Total:      1.73GB     1.73GB (flat, cum) 0.037%

github.com/thanos-io/thanos/pkg/receive.(*Handler).fanoutForward

/home/circleci/project/pkg/receive/handler.go

  Total:      1.68GB    11.44GB (flat, cum)  0.24%

google.golang.org/grpc/internal/transport.(*http2Server).WriteStatus

/home/circleci/go/pkg/mod/google.golang.org/grpc@v1.29.1/internal/transport/http2_server.go

  Total:      1.64GB     3.55GB (flat, cum) 0.075%

google.golang.org/grpc/internal/transport.(*http2Server).writeHeaderLocked

/home/circleci/go/pkg/mod/google.golang.org/grpc@v1.29.1/internal/transport/http2_server.go

  Total:      1.63GB     1.91GB (flat, cum)  0.04%

github.com/prometheus/prometheus/tsdb.(*blockBaseSeriesSet).Next.func1

/home/circleci/go/pkg/mod/github.com/prometheus/prometheus@v1.8.2-0.20201119142752-3ad25a6dc3d9/tsdb/querier.go

  Total:      1.60GB     1.60GB (flat, cum) 0.034%

github.com/grpc-ecosystem/go-grpc-middleware/tracing/opentracing.newClientSpanFromContext

/home/circleci/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.1.0/tracing/opentracing/client_interceptors.go

  Total:      1.56GB     3.32GB (flat, cum)  0.07%

google.golang.org/grpc/internal/transport.(*recvBuffer).put

/home/circleci/go/pkg/mod/google.golang.org/grpc@v1.29.1/internal/transport/transport.go

  Total:      1.44GB     1.44GB (flat, cum)  0.03%

google.golang.org/grpc/internal/transport.(*http2Client).NewStream

/home/circleci/go/pkg/mod/google.golang.org/grpc@v1.29.1/internal/transport/http2_client.go

  Total:      1.40GB    21.53GB (flat, cum)  0.46%

github.com/prometheus/prometheus/tsdb.seriesHashmap.set

/home/circleci/go/pkg/mod/github.com/prometheus/prometheus@v1.8.2-0.20201119142752-3ad25a6dc3d9/tsdb/head.go

  Total:      1.35GB     1.35GB (flat, cum) 0.029%

golang.org/x/net/http2.parseHeadersFrame

/home/circleci/go/pkg/mod/golang.org/x/net@v0.0.0-20201110031124-69a78807bb2b/http2/frame.go

  Total:      1.35GB     1.35GB (flat, cum) 0.029%

google.golang.org/grpc.(*Server).processUnaryRPC

/home/circleci/go/pkg/mod/google.golang.org/grpc@v1.29.1/server.go

  Total:      1.23GB     2.28TB (flat, cum) 49.55%

google.golang.org/grpc.(*clientStream).newAttemptLocked

/home/circleci/go/pkg/mod/google.golang.org/grpc@v1.29.1/stream.go

  Total:      1.18GB     1.18GB (flat, cum) 0.025%

```

**Comment by jmichalek132** [comment]
I also observed this issue with image: `thanosio/thanos:master-2021-01-05-0aa07118`.

**Comment by bwplotka** [comment]
BTW those memory metrics you have is for what exactly metrics? (it matters, before Go1.6)

**Comment by bwplotka** [comment]
> container_memory_usage_bytes

is inflated, see https://www.bwplotka.dev/2019/golang-memory-monitoring/

Some amazing ideas from Cortex experience are using following options (adding those envvars to Thanos process): 

* `GODEBUG=madvdontneed=1` which will mimick what Go1.16 will do by default (and what was done before Go1.12). This will improve observability part of things.
* `GOGC=50`  (default is 100). This changes GC to run more often. For memory heavy containers we want that.

**Comment by bwplotka** [comment]
While ingesting ~10 millions series (replicated) on 0.19.0 looks actually better not worse:

![image](https://user-images.githubusercontent.com/6950331/110766741-6693ff80-824d-11eb-8f8a-3640824daa04.png)

**Comment by jmichalek132** [comment]
![Screenshot 2021-03-11 at 16 33 37](https://user-images.githubusercontent.com/7171250/110812562-01a9cb00-8288-11eb-9a59-14da34fbf599.png)
Tested version in order
1. v0.17.2
2. v0.16.0
3. v0.19.0-rc.0
4. 1fff9a7157d1f3b7d5e0ce36f999b10bb3ae7ec4
5. c534b6d4a2f9fdac572f3126adfc1fb3e1b17147
6. b4528880101a5544aaab0c0cfd2f3d8bb9c22b80
7. 98753400f8885e57aa8958cdedccaa9c405e5a58
8. f494a9917413941feee96f577ead707d97d6cce8
9. v0.16.0
10. 2e128408d9a03579c73665ea469b82e380a0403d

**Comment by jmichalek132** [comment]
I am not sure that this [bugfix](https://github.com/thanos-io/thanos/commit/2e128408d9a03579c73665ea469b82e380a0403d) that was merged into 0.16.0 release branch made it back into main branch.

**Comment by metalmatze** [comment]
@jmichalek132 I think you're right. I was never merged back into v0.17+ but instead entirely replaced with the ZLabel.
https://github.com/thanos-io/thanos/commits/v0.17.0/pkg/store/labelpb/label.go

I'm wondering if @bwplotka based that work on the fixed label or ignored it and did an entire rewrite?
Either way, it shouldn't show up anymore with the benchmarks posted, it seems rather unlikely. :thinking:

**Comment by brancz** [comment]
If anyone on this thread would like to we'd be happy to give you access to our Polar Signals continuous profiler beta so we can continuously profile your thanos infrastructure to find the culprit! Just join our discord and we'll get you setup! https://github.com/thanos-io/thanos/pull/3919

**Comment by bwplotka** [comment]
What's the latest? 🤗

**Comment by bwplotka** [comment]
We discussed some of this today's Contributor Hours: https://docs.google.com/document/d/137XnxfOT2p1NcNUq6NWZjwmtlSdA6Wyti86Pd6cyQhs/edit#heading=h.dmnpchivqkn9 I will try to look on this a bit.

**Comment by bwplotka** [comment]
Investigation: I found this https://github.com/thanos-io/thanos/pull/3327 being on v0.16.0 but not on master. I think merge to v0.16.0 failed.

The long term fix was merged, but it looks like not properly, as this issue is exactly the same as https://github.com/thanos-io/thanos/issues/3265

Long term fix attempt no 1: https://github.com/thanos-io/thanos/pull/3279
Long "fix" merged: https://github.com/thanos-io/thanos/pull/3330 might be not actually fixing this. Let me unpack this (:

**Comment by bwplotka** [comment]
Thank you all for your patience and help. It kind of silly but the fix was already in PR but never merged: https://github.com/thanos-io/thanos/pull/3334/

Up-to-date fix is available here: https://github.com/thanos-io/thanos/pull/3943 and will be part of v0.19.0 🚀 

Some learnings when we would be investigating such issues:
* Find the regression commit. I found that only after this guy things were working fine for @jmichalek132 

![image](https://user-images.githubusercontent.com/6950331/111871997-e411eb00-8984-11eb-8116-c59b46d49487.png)

So it was easy to tell that such "quick fix" never made to master 🤗 

So it was as easy as really porting https://github.com/thanos-io/thanos/pull/3334/ (stripping from unrelated changes) and ensure we have profiles that back up our thinking. E.g those:

Before fix: https://share.polarsignals.com/68255aa/
After fix: https://share.polarsignals.com/fbffd26/

Notice this big 200MB chunk that does not exist now 🤗

**Comment by luizrojo** [comment]
Awesome news @bwplotka !

Looking forward to v0.19

**Comment by mxmorin** [comment]
I've tested 0.19-rc2 and issued is fixed.
Many thanks

**Comment by bwplotka** [comment]
Done then. I see the capacity to reduce the resource usage after v0.19.0 a lot too. Let's release it and iterate. Thanks all for help.
